hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractCompletableFuture.java

```diff
@@ -110,6 +167,32 @@ public V get() throws InterruptedException, ExecutionException {
         }
     }
 
+    @Override
+    public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+        final long deadlineNanoTime = System.nanoTime() + unit.toNanos(timeout);
+        long nanosToWait;
+        for (; ; ) {
+            if (isCancelled()) {
+                throw new CancellationException();
+            }
+            if (isDone()) {
+                return getResult();
+            }
+            if (Thread.interrupted()) {
+                throw new InterruptedException();
+            }
+
+            nanosToWait = deadlineNanoTime - System.nanoTime();
+            if (nanosToWait < 0) {
+                throw new TimeoutException();
+            }
+
+            synchronized (this) {
```

Comment:
Imagine 2 threads: 

- Thread1 - Waiter
- Thread2 - Producer

Waiter calls `get()`. Result is still not set so the Waiter will fall through up to line 190. But before Waiter enters the synchronized block the Producer thread will set the result. The Producer will call `notifyThreadsWaitingOnGet()`, but there is no thread waiting! (as the Waiter is right *before* the synchronized block)

Eventually Waiter reaches the synchronized block and calls `wait()`, but there is no one to call notify! (as the Producer is already gone). Did I miss something?